android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiUIWebView.java

```diff
@@ -491,6 +495,8 @@ public void propertyChanged(String key, Object oldValue, Object newValue, KrollP
 			}
 		} else if (TiC.PROPERTY_DISABLE_CONTEXT_MENU.equals(key)) {
 			disableContextMenu = TiConvert.toBoolean(newValue);
+		} else if (TiC.PROPERTY_ZOOM_LEVEL.equals(key)) {
+			getWebView().zoomBy(TiConvert.toFloat(newValue));
```

Comment:
The `WebView.zoomBy()` method is only available on API Level 21 (Android 5.0) and higher. So, calling this method on older Android OS versions will cause a crash.
https://developer.android.com/reference/android/webkit/WebView.html#zoomBy(float)

We also need to constrain the zoom level between 0.1 to 100.0 to avoid an IllegalArgumentException.
https://github.com/android/platform_frameworks_base/blob/master/core/java/android/webkit/WebView.java#L2111

I recommend that you guard the higher API level call as shown below. Putting the higher API level call within a private inner Java class avoids the opcode warning and is how Google does it via their support libraries. Read Google's blog post [here](https://android-developers.googleblog.com/2010/07/how-to-have-your-cupcake-and-eat-it-too.html) for details on this concept.

```java
} else if (TiC.PROPERTY_ZOOM_LEVEL.equals(key)) {
	if (Build.VERSION.SDK_INT >= 21) {
		ApiLevel21.zoomBy(getWebView(), TiConvert.toFloat(newValue, 1.0f));
	}
}


private static class ApiLevel21
{
	private ApiLevel21() {}

	public void zoomBy(WebView webView, float scale)
	{
		if (webView != null) {
			if (scale < 1.0f) {
				scale = 1.0f;
			} else if (scale >= 100.0f) {
				scale = 100.0f;
			}
			webView.zoomBy(scale);
		}
	}
}
```

The only bummer here is with Android 4.x which does not support the zoomBy() method. The older OS versions only support zoomIn() and zoomOut() methods.

Reply:
I don't think we benefit much from adding these inner classes to prevent logcat warnings on lower Android versions. Not all support classes do this either:
```
I/dalvikvm: Could not find method android.widget.FrameLayout.startActionModeForChild, referenced from method android.support.v7.widget.ActionBarContainer.startActionModeForChild
W/dalvikvm: VFY: unable to resolve virtual method 21612: Landroid/widget/FrameLayout;.startActionModeForChild (Landroid/view/View;Landroid/view/ActionMode$Callback;I)Landroid/view/ActionMode;
D/dalvikvm: VFY: replacing opcode 0x6f at 0x0002
W/dalvikvm: VFY: unable to find class referenced in signature (Landroid/graphics/drawable/Icon;)
I/dalvikvm: Could not find method android.widget.ImageButton.setImageIcon, referenced from method android.support.v7.widget.AppCompatImageButton.setImageIcon
W/dalvikvm: VFY: unable to resolve virtual method 21646: Landroid/widget/ImageButton;.setImageIcon (Landroid/graphics/drawable/Icon;)V
D/dalvikvm: VFY: replacing opcode 0x6f at 0x0000
E/dalvikvm: Could not find class 'android.graphics.drawable.RippleDrawable', referenced from method android.support.v7.widget.AppCompatImageHelper.hasOverlappingRendering
/dalvikvm: VFY: unable to resolve instanceof 205 (Landroid/graphics/drawable/RippleDrawable;) in Landroid/support/v7/widget/AppCompatImageHelper;
D/dalvikvm: VFY: replacing opcode 0x20 at 0x000c
```

Reply:
👍 I'll have a look at this